src/functional/ffront/type_info.py

```diff
@@ -1,48 +1,235 @@
-from dataclasses import dataclass
-from typing import Iterator, Optional, Type, TypeGuard, cast
+import enum
+from typing import Iterator, Type, TypeGuard
 
-from functional.common import GTTypeError
-from functional.ffront.common_types import (
-    DataType,
-    DeferredSymbolType,
-    FieldType,
-    FunctionType,
-    ScalarKind,
-    ScalarType,
-    SymbolType,
-    TupleType,
-    VoidType,
-)
+from functional.common import Dimension, GTTypeError
+from functional.ffront import common_types as ct
 
 
-@dataclass(frozen=True)
-class GenericDimensions:
-    """Sentry for generic dimensions to be returned instead of `Ellipsis`."""
-
-    def __len__(self):
-        return 0
-
-
-def is_complete_symbol_type(sym_type: Optional[SymbolType]) -> TypeGuard[SymbolType]:
+def is_concrete(symbol_type: ct.SymbolType) -> TypeGuard[ct.SymbolType]:
     """Figure out if the foast type is completely deduced."""
-    match sym_type:
-        case None:
-            return False
-        case DeferredSymbolType():
+    match symbol_type:
+        case ct.DeferredSymbolType():
             return False
-        case SymbolType():
+        case ct.SymbolType():
             return True
     return False
 
 
-def is_complete_scalar_type(sym_type: Optional[SymbolType]) -> TypeGuard[ScalarType]:
-    if isinstance(sym_type, ScalarType):
+class TypeKind(enum.Enum):
+    FIELD = 0
+    SCALAR = 1
+    UNKNOWN = 2
+
+
+def type_kind(symbol_type: ct.SymbolType) -> TypeKind:
```

Comment:
Deducing if something is a field or a scalar is a common thing to ask, but can you make a case for the behavior of this function that goes beyond what can be done with a `type_class(symbol_type) == ct.FieldType | ct.ScalarKind`? I would argue that the additional behavior creates more harm then good. A tuple of fields or a function returning a field behaves different to a field not only with respect to the entity itself, but also with respect to the information the `SymbolType` will provide to you. Arguably they should also be treated differently. Consider for example https://github.com/GridTools/gt4py/blob/3b0425deee343ed890377cd9f396d40cdd746733/src/functional/ffront/type_info.py#L225-L227 which does not make sense for a function returning fields, but is still relying on this function. If you want to avoid writing out the `type_class` comparison each and every time I would propose to introduce two functions `is_field` and `is_scalar`.

Reply:
See above for the general reasoning. The example here is definitely a place where I should have used `type_class` instead of `type_kind`.

Consider the following usage though:
 https://github.com/GridTools/gt4py/blob/3b0425deee343ed890377cd9f396d40cdd746733/src/functional/ffront/func_to_foast.py#L222

If you change it to `elif type_class(...) is ct.ScalarType` it will fail because the question we should be asking there is not "Is it a scalar?" but "Is the left hand side of the assignment going to be a scalar?", which is equivalent to "Is the resulting type kind scalar?".

Another motivating example is of course the lowering, where we want to know: "Should we treat this as a 'Value' or as an 'Iterator' (or do the concepts not apply)?". Which turns out to be equivalent to "Is the resulting type kind 'Scalar', 'Field' or 'Unknown'"?

Reply:
To make it even less confusing, I have replaced all usage except in the lowering by `type_info.is_concrete(...) and type_info.type_kind(...) is ScalarType | FieldType`. I am not sure that is the right question to ask, but all the tests pass, so it seems a good enough question to ask for now.